osquery/tables/system/windows/ntfs_acl_permissions.cpp

```diff
@@ -0,0 +1,207 @@
+#include <Aclapi.h>
+#include <map>
+#include <string>
+#include <vector>
+#include <windows.h>
+
+#include <boost/algorithm/string/join.hpp>
+#include <boost/filesystem.hpp>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+namespace alg = boost::algorithm;
+namespace fs = boost::filesystem;
+
+namespace osquery {
+namespace tables {
+
+// helper function to get access mode string
+std::string accessModeToStr(const _ACCESS_MODE& accessMode) {
+  switch (accessMode) {
+  case NOT_USED_ACCESS:
+    return "Not Used";
+  case GRANT_ACCESS:
+    return "Grant";
+  case SET_ACCESS:
+    return "Set";
+  case DENY_ACCESS:
+    return "Deny";
+  case REVOKE_ACCESS:
+    return "Revoke";
+  case SET_AUDIT_SUCCESS:
+    return "Set Audit Success";
+  case SET_AUDIT_FAILURE:
+    return "Set Audit Failure";
+  }
+  return "";
+}
+
+// helper function to build access string from permission bit mask
+std::string accessPermsToStr(const unsigned long pmask) {
+  std::vector<std::string> permList;
+  const std::map<unsigned long, std::string> permVals = {
+      {DELETE, "Delete"},
+      {READ_CONTROL, "Read Control"},
+      {WRITE_DAC, "Write DAC"},
+      {WRITE_OWNER, "Write Owner"},
+      {SYNCHRONIZE, "Synchronize"},
+      {STANDARD_RIGHTS_REQUIRED, "Std Rights Required"},
+      {STANDARD_RIGHTS_ALL, "Std Rights All"},
+      {SPECIFIC_RIGHTS_ALL, "Specific Rights All"},
+      {ACCESS_SYSTEM_SECURITY, "Access System Security"},
+      {MAXIMUM_ALLOWED, "Maximum Allowed"},
+      {GENERIC_READ, "Generic Read"},
+      {GENERIC_WRITE, "Generic Write"},
+      {GENERIC_EXECUTE, "Generic Execute"},
+      {GENERIC_ALL, "Generic All"}};
+
+  for (auto const& perm : permVals) {
+    if ((pmask & perm.first) != 0) {
+      permList.push_back(perm.second);
+    }
+  }
+
+  return alg::join(permList, ",");
+}
+
+// helper function to convert inheritance type to string
+std::string inheritanceToStr(const unsigned long i) {
+  switch (i) {
+  case CONTAINER_INHERIT_ACE: // equivalent to SUB_CONTAINERS_ONLY_INHERIT
+    return "Container Inherit Ace";
+  case INHERIT_NO_PROPAGATE: // equivalent to NO_PROPAGATE_INHERIT_ACE
+    return "Inherit No Propagate";
+  case INHERIT_ONLY: // equivalent to INHERIT_ONLY_ACE
+    return "Inherit Only";
+  case OBJECT_INHERIT_ACE: // equivalent to SUB_OBJECTS_ONLY_INHERIT
+    return "Object Inherit Ace";
+  case SUB_CONTAINERS_AND_OBJECTS_INHERIT:
+    return "Sub containers and Objects Inherit";
+  case NO_INHERITANCE:
+    return "No Inheritence";
+  default:
+    return "Other";
+  }
+}
+
+// helper function to get account/group name from trustee
+std::string trusteeToStr(const TRUSTEE& t) {
+  char name[256];
+  char domain[256];
+  unsigned long size = 256; // Max username length for Windows
+  SID_NAME_USE accountType;
+
+  switch (t.TrusteeForm) {
+  case TRUSTEE_IS_SID:
+    // get the name from the SID
+    if (!LookupAccountSid(nullptr,
+                          reinterpret_cast<PSID>(t.ptstrName),
```

Comment:
Make every effort to avoid `reinterpret_cast` if possible. In this case, are you sure that you can't initialize with a `static_cast` or a locally initialized variable?

Reply:
I took out as many of the reinterpret_casts as I could but the place they remain are required 😢